### PR TITLE
Introduce noRetry Parameter for checkcapacity ProvisioningRequest

### LINKED ISSUE #7495: Add Parameter to checkcapacity ProvisioningRequest that disables CA's retry mechanism
**Which component are you using?**:
Cluster-autoscaler

**Is your feature request designed to solve a problem? If so describe the problem this feature should solve.**:

If the checkcapacity ProvisioningRequest cannot be provisioned, it waits for backoff time before being retried. From the perspective of integration with Kueue, this time is wasteful as we block quota on Kueue's side, even it's not actively used. 

**Describe the solution you'd like.**:
Add ProvisioningRequest Parameter that disables retry mechanism in CA. If the Parameter is present and the request cannot be provisioned it should be marked as `Failed=True` instead of `Provisioned=False`

**Describe any alternative solutions you've considered.**:
Add flag to the CA that disables retry mechanism. However we need more granular solution, because of different nature of some workloads - depending on a workload users are willing to block(or not) quota

**Additional context.**:

### PR DESCRIPTION
**Pull request by PBundyra**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Introduces a new Parameter for checkcapacity ProvisioningRequests that limits retry mechanism in ClusterAutoscaler.
If the capacity can't be found then the ProvisioningRequest is marked as `Failed=True` instead of `Provisioned=False`

#### Which issue(s) this PR fixes:

Fixes #7495 

#### Does this PR introduce a user-facing change?

```release-note
New Parameter for checkcapacity ProvisioningRequests that limits retry mechanism
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by PBundyra** [comment]
/assign @yaroslava-serdiuk 
cc @mwielgus

**Comment by yaroslava-serdiuk** [comment]
Could you add a test case for your change? The majority of cases live in orchestrator_test file.

File: cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go [hunk 1]
```diff
@@ -139,7 +143,11 @@ func (o *checkCapacityProvClass) checkcapacity(unschedulablePods []*apiv1.Pod, p
 	err, cleanupErr := clustersnapshot.WithForkedSnapshot(o.context.ClusterSnapshot, func() (bool, error) {
 		st, _, err := o.schedulingSimulator.TrySchedulePods(o.context.ClusterSnapshot, unschedulablePods, scheduling.ScheduleAnywhere, true)
 		if len(st) < len(unschedulablePods) || err != nil {
-			conditions.AddOrUpdateCondition(provReq, v1.Provisioned, metav1.ConditionFalse, conditions.CapacityIsNotFoundReason, "Capacity is not found, CA will try to find it later.", metav1.Now())
+			if _, ok := provReq.Spec.Parameters[NoRetryParameter]; ok {
```
Comment (yaroslava-serdiuk) [inline cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go:146]: Could you add a command which will explain the use case and the reasoning why noRetry parameter leads to failed condition, please?

Comment (aleksandra-malinowska) [inline cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go:146]: Do we really want to assume any value for "noRetry" means "true"?

Even if user inputs "false"?

Comment (PBundyra) [reply cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go]: Fixed

Comment (PBundyra) [reply cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go]: Sure, done

**Comment by PBundyra** [comment]
> Could you add a test case for your change? The majority of cases live in orchestrator_test file.

Thanks, added

File: cluster-autoscaler/provisioningrequest/provreqwrapper/wrapper.go [hunk 2]
```diff
@@ -81,3 +81,15 @@ func errMissingPodTemplates(podSets []v1.PodSet, podTemplates []*apiv1.PodTempla
 	}
 	return fmt.Errorf("missing pod templates, %d pod templates were referenced, %d templates were missing: %s", len(podSets), len(missingTemplates), strings.Join(missingTemplates, ","))
 }
+
+// Parameters makes a deep copy of embedded ProvReq and sets its Parameters
+func (pr *ProvisioningRequest) Parameters(params map[string]v1.Parameter) *ProvisioningRequest {
+	prCopy := pr.DeepCopy()
```
Comment (yaroslava-serdiuk) [inline cluster-autoscaler/provisioningrequest/provreqwrapper/wrapper.go:87]: It's a test utility function, so let's move it to the testutils file.

File: cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go [hunk 3]
```diff
@@ -139,7 +143,13 @@ func (o *checkCapacityProvClass) checkcapacity(unschedulablePods []*apiv1.Pod, p
 	err, cleanupErr := clustersnapshot.WithForkedSnapshot(o.context.ClusterSnapshot, func() (bool, error) {
 		st, _, err := o.schedulingSimulator.TrySchedulePods(o.context.ClusterSnapshot, unschedulablePods, scheduling.ScheduleAnywhere, true)
 		if len(st) < len(unschedulablePods) || err != nil {
-			conditions.AddOrUpdateCondition(provReq, v1.Provisioned, metav1.ConditionFalse, conditions.CapacityIsNotFoundReason, "Capacity is not found, CA will try to find it later.", metav1.Now())
+			if noRetry, ok := provReq.Spec.Parameters[NoRetryParameter]; ok && noRetry == "true" {
```
Comment (aleksandra-malinowska) [inline cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go:146]: Bordering on nit: it isn't great experience if invalid value for a known parameter is just quietly ignored.

I understand if you don't want to extend the scope of this CL, but I think for better UX in the future CA should inform the user that it's ignoring a supported parameter (at minimum by logging a warning that value is invalid), or better yet, fail the request with similar error.

Comment (PBundyra) [reply cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go:146]: I agree, for this parameter I'll log if the value is invalid (the only valid value is `true` actually) however I believe CA should have more general mechanism for validating parameters. Hence I'll create an issue for that

File: cluster-autoscaler/provisioningrequest/provreqwrapper/wrapper.go (hunk 2 above)
Comment (PBundyra) [reply cluster-autoscaler/provisioningrequest/provreqwrapper/wrapper.go]: Done

File: cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go (hunk 3 above)
Comment (PBundyra) [reply cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go]: Thanks for pointing this out

**Comment by aleksandra-malinowska** [comment]
/lgtm
/approve
/hold

Couple of nits only, otherwise LGTM. Feel free to remove hold when you consider this ready.

File: cluster-autoscaler/provisioningrequest/orchestrator/orchestrator_test.go [hunk 4]
```diff
@@ -242,6 +243,15 @@ func TestScaleUp(t *testing.T) {
 			provReqToScaleUp: newCheckCapacityCpuProvReq,
 			scaleUpResult:    status.ScaleUpSuccessful,
 		},
+		{
+			name: "impossible check-capacity, with noRetry parameter",
+			provReqs: []*provreqwrapper.ProvisioningRequest{
+				impossibleCheckCapacityReq.Parameters(map[string]v1.Parameter{"noRetry": "true"}),
```
[the remark this answers is not in the document]
Comment (PBundyra) [reply cluster-autoscaler/provisioningrequest/orchestrator/orchestrator_test.go]: Good point, thanks

File: cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go [hunk 5]
```diff
@@ -139,7 +147,16 @@ func (o *checkCapacityProvClass) checkcapacity(unschedulablePods []*apiv1.Pod, p
...
 		st, _, err := o.schedulingSimulator.TrySchedulePods(o.context.ClusterSnapshot, unschedulablePods, scheduling.ScheduleAnywhere, true)
 		if len(st) < len(unschedulablePods) || err != nil {
-			conditions.AddOrUpdateCondition(provReq, v1.Provisioned, metav1.ConditionFalse, conditions.CapacityIsNotFoundReason, "Capacity is not found, CA will try to find it later.", metav1.Now())
+			if noRetry, ok := provReq.Spec.Parameters[NoRetryParameterKey]; ok && noRetry == "true" {
+				// Failed=true condition triggers retry in Kueue. Otherwise ProvisioningRequest with Provisioned=Failed
+				// condition block capacity in Kueue even it's in the middle of backoff waiting time.
+				conditions.AddOrUpdateCondition(provReq, v1.Failed, metav1.ConditionTrue, conditions.CapacityIsNotFoundReason, "CA could not find requested capacity", metav1.Now())
+			} else {
+				if noRetry, ok := provReq.Spec.Parameters[NoRetryParameterKey]; ok && noRetry != "false" {
+					klog.Errorf("Invalid value: %v for %v Parameter in %v ProvisioningRequest. Supported values are: \"true\", \"false\"", noRetry, NoRetryParameterKey, provReq.Name)
```
[the remark this answers is not in the document]
Comment (PBundyra) [reply cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go]: Done

File: cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go [hunk 6]
```diff
@@ -139,7 +147,16 @@ func (o *checkCapacityProvClass) checkcapacity(unschedulablePods []*apiv1.Pod, p
 	err, cleanupErr := clustersnapshot.WithForkedSnapshot(o.context.ClusterSnapshot, func() (bool, error) {
 		st, _, err := o.schedulingSimulator.TrySchedulePods(o.context.ClusterSnapshot, unschedulablePods, scheduling.ScheduleAnywhere, true)
 		if len(st) < len(unschedulablePods) || err != nil {
-			conditions.AddOrUpdateCondition(provReq, v1.Provisioned, metav1.ConditionFalse, conditions.CapacityIsNotFoundReason, "Capacity is not found, CA will try to find it later.", metav1.Now())
+			if noRetry, ok := provReq.Spec.Parameters[NoRetryParameterKey]; ok && noRetry == "true" {
+				// Failed=true condition triggers retry in Kueue. Otherwise ProvisioningRequest with Provisioned=Failed
+				// condition block capacity in Kueue even it's in the middle of backoff waiting time.
```
[the remark this answers is not in the document]
Comment (PBundyra) [reply cluster-autoscaler/provisioningrequest/checkcapacity/provisioningclass.go]: Fixed

**Comment by PBundyra** [comment]
Thanks for reviews
/hold cancel

**Pull request merged**